### DISCUSSION TITLE
New consignment format

### DISCUSSION THREAD

**Discussion by dr-orlovsky** [RGB-WG/rgb#57] [Q&A]
Consignments are blobs of client-side-validated data which has to be sent between users in order for RGB state transfers to happen.

Each consignment must include all information required for its validation - except information about bitcoin transactions involved in the contract (this info is provided by a bitcoin core instance).

Such information includes:
- schema;
- contract genesis;
- set of strict type libraries (providing information about data types used by the contract);
- set of AluVM libraries running validation programs;
- collection of anchors with corresponding state transition bundles which may contain partially concealed information (_NB: data attachments for state transitions are **not** a part of a consignment!_);
- set of state extensions used by any of state transitions.

Alongside these consensus data consignment must also provide the following meta information:
- contract manifest which provides information about used RGB interface standards may include different media metadata;
- certificate and signature over schema for the schema developer;
- certificate and signature over genesis and manifest for the contract issuer;
- optional opentimestamp proof for the genesis.

Consignments may be small - just several kbytes - and grow quite large up to multiple gigabytes in size. They can be distributes via different channels, starting with dynamic QR codes (consignments with issued assets), instant messages, e-mail - or as files via FTP and HTTP. For small consignments a Base64 ascii armoring is provided to support distribution via QRs and messages. For large consignments sent over HTTP a streaming mode should be available such that the receiving party may run validation in parallel with downloading - and abort if the consignment is invalid without the need to download multiple gigabytes of data (this should prevent some of DoS attacks). 

Thus, consignment binary data must support different forms:
- base64 armored text block;
- single binary file which must support simple seeking to the desired data piece (see the list of consignment components above);
- FS package (directory or package for OSes like MacOS);
- streamed data.

Content of the consignment is highly entropic (due to txids, Petersen commitments, bulletproofs, anchor data being >80% of the most of the consignments) - thus, application of additional compression is not desired.

Here I propose to discuss which format must be the best to work with such requirements.

**Comment by cryptoquick** [comment]
Compression does help rgbc consignment files, from my experience, it's one of Carbonado's test files:
https://github.com/diba-io/carbonado/tree/main/tests/samples
For FS, Carbonado could be a really good container format, since it can be encrypted on-disk using wallet keys, and can be handed off to other hosts to be verified remotely. It also builds in optional error correction on a binary level (not like bech32 bch).

**Reply by dr-orlovsky** [reply]
We discussed using Carbonado and you confused me a bit saying it is not a good fit:
- why would we need error correction?
- why we need to mix encryption with storage? Why not to encrypt the file as a whole?
- how we can support streamed verification mode?
- when we have consignment as an on-disk file can we seek to a certain data component in a simple way?

**Reply by cryptoquick** [reply]
It's good for consignment encoding, it's one of the major reasons why we built it, as described in the README.

- In case it's not handled by the storage media, and if it is, that's fine, it's optional, all the encoding steps are. Error correction of some kind is important for consignment data because if it's ever lost or corrupted, the tokens are effectively lost.
- This is file encryption, yes. Databases don't necessarily include or support that.
- Stream verification is something that can be done by a separate node to make periodic payments to those holding your consignment data.
- Seeking would require a special io::Read/Write trait implementation, but definitely doable if provided a byte offset. An index could be kept separately.

**Reply by dr-orlovsky** [reply]
Well, then I got you wrong - it seems on our call I directly asked about applying Carbonado for consignment and I recall you saying it is probably not suited for it...

Anyway, two things:

> - Stream verification is something that can be done by a separate node to make periodic payments to those holding your consignment data.

I meant a different thing: can a carbonado-encoded data be sent via HTTP streaming such that my parser will be able to parse the data on-the-fly into specific components (like "here one state transition ends and another one begins)?

Also, if the streaming is broken, can I request to download from a specific point? (I assume yes, it is more implementation thing rather then file format).

> - Seeking would require a special io::Read/Write trait implementation, but definitely doable if provided a byte offset. An index could be kept separately.

I mean does Carbonado provides some sort of header with offsets to all data components for a multi-component files (like consignments are)?

**Reply by dr-orlovsky** [reply]
So my take was that Carbonado is designed for storing stash backups, not consignments - consignments are temporary and do not persist at all, thus they do not need a mayday protection (a sender can create a new consignment as long as it is needed).

**Reply by cryptoquick** [reply]
> Also, if the streaming is broken, can I request to download from a specific point? (I assume yes, it is more implementation thing rather then file format).

Yes, picking up from a byte offset would be something that we'd like to support on HTTP frontends for the carbonado-node.

> I mean does Carbonado provides some sort of header with offsets to all data components for a multi-component files (like consignments are)?

That's planned with Catalogs, it's mentioned in the README, and there's also an issue:
https://github.com/diba-io/carbonado/issues/3

> So my take was that Carbonado is designed for storing stash backups, not consignments - consignments are temporary and do not persist at all, thus they do not need a mayday protection (a sender can create a new consignment as long as it is needed).

The RGB methods in bitmask-core are stateless, so a consignment is effectively the same as a stash. They can also be encrypted with ECDH keys, so they can be privately shared with peers. This helps if an insecure or unauthenticated means of communication were used, such as email.

**Reply by dr-orlovsky** [reply]
Consignment can't be a stash since stash usually covers multiple contracts, while consignments are always single-contract specific. Stash can be exported only in form of a disclosure, but this is also a highly suboptimal.

**Reply by cryptoquick** [reply]
Yes, we export them and persist them separately.

**Reply by cryptoquick** [reply]
Also, when multiple assets are transferred in a single transaction, are multiple consignments produced?

**Reply by dr-orlovsky** [reply]
In RGB there are no "transactions", so it is not clear what do you mean "assets transfered by a single transaction".

If you mean a single witness transaction, than yes, it may have many RGB and non-RGB data for tapret commitments, but it has one-to-many relations with consignments.
